tls/tls_ticket.c

```diff
@@ -22,409 +22,646 @@
+#define TTLS_TICKETS_TAG_LEN		8
+
+/* ---- Configuration and Key management                                 ---- */
 
 /*
- * Generate/update a key
- * TODO #1054 use a configuration option to generate the key with the random
- * number generator or on top of shared secret.
+ * Initialisation vector for session ticket master key. Hard-coded to allow
+ * the same key generation on all Tempesta nodes with the same user
+ * configuration and user secrets. It's ok, that attacker may know it,
+ * it's just a value HMAC'ed with a really secret key.
  */
-static int ssl_ticket_gen_key(ttls_ticket_context *ctx,
-				   unsigned char index)
+const char *ticket_secret_key_iv =
+	"u5xBNXmcQwxs9yGfv3IJa0h3QIZujnuf0ISmycYSB4vhfitCMM1phNP9ft3xjEbR";
+/* for ticket symmetric key: */
+const char *ticket_key_sym_iv =
+	"r26bMJcfLdlYyn9wM3xsHrzraeLKQHGCgYkWivTu6UVxw7VxcJQAr63k8Sa6lFUa";
+/* for ticket key name: */
+const char *ticket_key_name_iv =
+	"qolUvqou29yxSwvz2jWTNvk3znIjy25E";
+
+#define SLEN(s)	(sizeof(s) - 1)
+
+static inline unsigned long
+ttls_ticket_get_time(unsigned long lifetime)
 {
-	int ret;
-	unsigned char buf[MAX_KEY_BYTES];
-	ttls_ticket_key *key = ctx->keys + index;
+	unsigned long ts = tfw_current_timestamp();
 
-	key->generation_time = (uint32_t) ttls_time();
+	ts -= ts % lifetime;
 
-	ttls_rnd(key->name, sizeof(key->name));
-	ttls_rnd(buf, sizeof(buf));
+	return ts;
+}
 
-	/* With GCM and CCM, same context can encrypt & decrypt */
-	ret = ttls_cipher_setkey(&key->ctx, buf, key->ctx.cipher_info->key_len,
-				 TTLS_ENCRYPT);
+static int
+__ttls_ticket_gen_key(TlsTicketKey *key, unsigned long ts,
+		      const char *secret)
+{
+	TlsMdCtx md_ctx;
+	int r;
 
-	bzero_fast(buf, sizeof(buf));
+	key->ts = ts;
+	ttls_md_init(&md_ctx);
+	if ((r = ttls_md_setup(&md_ctx, t_cfg.md_info, 1))) {
+		T_ERR_NL("TLS: can't init ");
+	}
+	r |= ttls_md_hmac_starts(&md_ctx, secret, TTLS_TICKET_KEY_LEN);
+	r |= ttls_md_hmac_update(&md_ctx, ticket_key_sym_iv,
+				 SLEN(ticket_key_sym_iv));
+	r |= ttls_md_hmac_update(&md_ctx, (unsigned char *)&key->ts,
+				 sizeof(key->ts));
+	r |= ttls_md_finish(&md_ctx, key->key);
+	ttls_md_free(&md_ctx);
+
+	/*
+	 * Set key->ts to 0 to indicate that the calculation has failed and
+	 * not try to use the key on every new handshake.
+	 */
+	if (r) {
+		T_WARN("TLS: can't rotate tls key");
+		key->ts = 0;
+		return r;
+	}
 
-	return ret;
+	return r;
 }
 
-/*
- * Rotate/generate keys if necessary
- */
-static int ssl_ticket_update_keys(ttls_ticket_context *ctx)
+static inline int
+ttls_ticket_gen_key(TlsTicketKey *key, const char *secret,  unsigned long ts)
 {
-	if (ctx->ticket_lifetime != 0)
-	{
-		uint32_t current_time = (uint32_t) ttls_time();
-		uint32_t key_time = ctx->keys[ctx->active].generation_time;
+	int r;
 
-		if (current_time > key_time &&
-			current_time - key_time < ctx->ticket_lifetime)
-		{
-			return 0;
-		}
+	write_lock(&key->lock);
+	r = __ttls_ticket_gen_key(key, ts, secret);
+	write_unlock(&key->lock);
```

Comment:
All the write locks are acquired in timer, i.e. there is only one writer and we don't have concurrent writers. But why readers can no do progress while we're rotating keys?

Reply:
The locking procedure was reviewed, now the write locks are acquired as little as possible.